bin/olly.ml

```diff
@@ -171,30 +190,38 @@ let trace format trace_filename exec_args =
       in
       let init () = () in
       let cleanup () = Trace.close trace_file in
-      olly ~runtime_begin ~runtime_end ~init ~cleanup exec_args
+      olly ~runtime_begin ~runtime_end ~init ~lifecycle ~cleanup exec_args
 
-let latency json output exec_args =
+let gc_stats json output exec_args =
   let current_event = Hashtbl.create 13 in
   let hist =
     H.init ~lowest_discernible_value:10 ~highest_trackable_value:10_000_000_000
       ~significant_figures:3
   in
+  let is_gc_phase phase =
+    match phase with
+    | "major" | "stw_leader" | "interrupt_remote" -> true
+    | _ -> false
+  in
   let runtime_begin ring_id ts phase =
     match Hashtbl.find_opt current_event ring_id with
     | None -> Hashtbl.add current_event ring_id (phase, Ts.to_int64 ts)
     | _ -> ()
   in
   let runtime_end ring_id ts phase =
     match Hashtbl.find_opt current_event ring_id with
-    | Some (saved_phase, saved_ts) when saved_phase = phase ->
+    | Some (saved_phase, saved_ts) when saved_phase = phase -> begin
         Hashtbl.remove current_event ring_id;
         let latency = Int64.to_int (Int64.sub (Ts.to_int64 ts) saved_ts) in
-        assert (H.record_value hist latency)
+        assert (H.record_value hist latency);
+        if is_gc_phase (Runtime_events.runtime_phase_name phase) then
+          Atomic.set total_gc_time (Atomic.get total_gc_time + latency)
```

Comment:
No need for Atomic here as olly itself is sequential. Either way the protocol is broken isn’t it? The get and set are individually atomic but the increment is not. You want increment operation of you want the entire thing to be atomic (but not needed here anyway).